### ISSUE TITLE
Support for Schema Registry?

### ISSUE THREAD

**Issue by cricket007** [Comcast/MirrorTool-for-Kafka-Connect#4]
Good to see an open-source version of Confluent Replicator floating around.  

I would like to ask about the scenario when a [Schema Registry has been deployed in two DC's](https://docs.confluent.io/current/schema-registry/docs/multidc.html). Here's the documentation on [Confluent Replicator](https://docs.confluent.io/current/multi-dc/index.html) about this setup. 

Not sure if Comcast is using a Schema Registry such as those offered by Hortonworks or Confluent, but I'll walk over the scenario when you alter the topic-name on the destination cluster. 

1. Topic `A` is created with `DC1` cluster. When produced to, a `KafkaAvroSerializer` class will create `A-key` and `A-value` "subjects" in the registry. 
2. Topic `A` is mirrored to `DC2` cluster as `aggregate.A`. 
3. The destination consumers using `KafkaAvroDeserializer` are now expecting to do a lookup against the registry for subjects `aggregate.A-key` and `aggregate.A-value`, 
4. The deserialization to do that lookup will fail (actually, the HTTP request to the registry) as this tool copied the Avro bytes, preserving the schema ID for topic `A`, but did not register those two new "subject" names in the destination registry. 

Mostly curious if anyone has tried to swap out `ByteArrayConverter` for [`AvroConverter`](https://github.com/confluentinc/schema-registry/blob/master/avro-converter/src/main/java/io/confluent/connect/avro/AvroConverter.java)

**Comment by rhysmccaig** [comment]
Hi Jordan,

Thanks for your question.

If I you understand correctly, in this scenario, the destination consumers are consuming from the `aggregate.*` topics, but the schema registry will not have topic->schema record as it only contains the entries for the original `*` topics?

I haven't personally tested using `AvroConverter` or any other converter in place of `ByteArrayConverter` as the connector itself uses/assumes byte arrays schema internally (it consumes key/messages as a byte array, and passes them to the internal Kafka connect producer as byte arrays).

Im not 100% sure how the internal producer works:
* if it first Deserializes the connect SourceRecords using the AvroConverter before re-serializing them when it produces to the designation cluster, it should just work to just change the `ByteArrayConverter` to `AvroConverter` in the configuration.
* If it just tries to serialize the objects it receives from the SourceConnector, as is, then this would fail, and I think that the best way to support this use case might be to implement an AvroTransformer class where the data flow would look like. (Rather than increase the complexity of the connector code)

`Source->Connector[ByteArray]->TransformToAvroRecord->Kafka Connect[AvroConverter]->Destination.`


On a somewhat related note - one limitation of this connector, as opposed to the confluent replicator, is that it doesn't allow you to fine tune the topic configuration on the destination cluster. (Topics must either be created in advance or they will be created with the default configuration if auto configuration is allowed.). If [KIP-158](https://cwiki.apache.org/confluence/display/KAFKA/KIP-158%3A+Kafka+Connect+should+allow+source+connectors+to+set+topic-specific+settings+for+new+topics) is implemented, I'd love to support that functionality

**Comment by cricket007** [comment]
> destination consumers are consuming from the `aggregate.*` topics, but the schema registry will not have topic->schema record as it only contains the entries for the original `*` topics?

That's correct. 

RE *uses/assumes byte arrays internally*  -- [so does Confluent's Avro format](https://docs.confluent.io/current/schema-registry/docs/serializer-formatter.html#wire-format), but that 1-4 byte slice needs to be extracted to get the schema ID for a particular message, or one could [use the `CachedSchemaRegistryClient` to directly fetch Schema metadata](https://github.com/confluentinc/schema-registry/blob/master/client/src/main/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClient.java#L242) by a given subject name (add `-key` or `-value` to the topic name). 

Those are the only two ways I know of that one would be able to extract schema to "sync" it to the destination registry. 

I am currently working for a client where we have extended MirrorMaker to support this functionality via a custom `MirrorMakerMessageHandler` implementation, but it actually only syncs schemas on calling our  constructor, not detects schema changes over time, but that is a future enhancement. 

Thinking on this a bit, I do like the idea of implementing a Connect SMT that would do that transformation/registration pipeline you mention rather than messing around with `toConnectData` / `fromConnectData` in the `Converter` classes, assuming `AvroConverter` doesn't work as expected.   

As for fine-tuning the destination topic - I've yet to use the [`AdminClient` API](http://kafka.apache.org/documentation/#adminclientconfigs) too much, but I believe it is the best option to read the source cluster's topic configuration without needing to hit Zookeeper, then using another instance of the class for the destination cluster, then use a combination of `AdminClient#createTopics` or `AdminClient#alterConfigs` depending if topic exists already.

**Comment by rhysmccaig** [comment]
> As for fine-tuning the destination topic - I've yet to use the AdminClient API too much, but I believe it is the best option to read the source cluster's topic configuration without needing to hit Zookeeper, then using another instance of the class for the destination cluster, then use a combination of AdminClient#createTopics or AdminClient#alterConfigs depending if topic exists already.

Yep - `AdminClient` would be the way to go - This connector uses the `AdminClient` to monitor the configured topic/partitions on the source cluster for changes to topic partitions to mirror. 

For now I don't want to add any code to the connector that deals with the destination cluster, outside of what is available in the connect API, as it feels a bit weird to have destination topic configuration in the source connector - specifically a code path which interacts with the destination. (Hence why i'm hoping KIP-158 is implemented for connect directly)

**Comment by cricket007** [comment]
Not a problem. I have finally gotten around to writing the SMT that I was thinking of. Will update if it works as expected. 

I had one question on this statement 

> it doesn't allow you to fine tune the topic configuration on the destination cluster

I did notice the use of the `TopicAdmin` client, which seems to at least carry over the partition count, haven't tried to see if topic configurations are carried over as well. Was that a change after that comment?

**Comment by rhysmccaig** [comment]
Would love to hear how you go with the SMT!

I haven't currently added any code to manage topic configs on the destination cluster - I do foresee needing this in the future, but in all honesty by the time I get around to it, KIP-382 might be ready to go.

**Comment by cricket007** [comment]
> ould love to hear how you go with the SMT!

Here ya go 🎉 
https://github.com/cricket007/schema-registry-transfer-smt

> I haven't currently added any code to manage topic configs on the destination cluster

Yeah, not sure if I dreamt that 😆  or was looking at another repo
